pkg/providers/ibmcloud.go

```diff
@@ -0,0 +1,331 @@
+	// would prevent a bunch of node not found errors in the logs from acquirecomplete
+
+	res.Id = id
+	return res, nil
+}
+
+func (p *ibmcloudProvider) getNodeByName(name string) (*datatypes.Hardware, error) {
+	service := services.GetAccountService(p.client)
+	nodes, err := service.Mask("id;hostname;primaryIpAddress;hardwareStatus;billingItem;hourlyBillingFlag").GetHardware()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, node := range nodes {
+		if *node.Hostname == name {
+			return &node, nil
+		}
+	}
+	return nil, errors.New("Node node found")
+}
+
+func (p *ibmcloudProvider) ensureSSHKey(newkey string) (*int, error) {
+	sshKeys, err := services.GetAccountService(p.client).GetSshKeys()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, key := range sshKeys {
+		if *key.Key == newkey {
+			return key.Id, nil
+		}
+	}
+	newKey := datatypes.Security_Ssh_Key{}
+	newKey.Key = sl.String(newkey)
+	// TODO: could do with something desciptive here (poolname if available)
+	newKey.Label = sl.String(strings.Split(uuid.New().String(), "-")[0])
+	newKey, err = services.GetSecuritySshKeyService(p.client).CreateObject(&newKey)
+
+	if err != nil {
+		return nil, err
+	}
+	return newKey.Id, nil
+}
+
+func (p *ibmcloudProvider) AcquireCompleted(id string) (bool, Resource, error) {
+
+	res := Resource{
+		Id: id,
+	}
+
+	node, err := p.getNodeByName(id)
+	if err != nil {
+		return false, res, err
+	}
+
+	// This may happen for a short period (20 seconds or so) after a node is initially created
+	if node == nil {
+		return false, res, fmt.Errorf("error getting node: %w", err)
+	}
+
+	// not ready yet
+	if *node.HardwareStatus.Status != "ACTIVE" {
+		return false, res, nil
+	}
+
+	res.Address = *node.PrimaryIpAddress
+
+	// Hold back on setting nodes to Available until ssh is available
+	if !utils.IsPortOpen(res.Address, "22") {
+		return false, res, nil
+	}
+	return true, res, nil
+}
+
+func (p *ibmcloudProvider) Clean(id string) error {
+	node, err := p.getNodeByName(id)
+
+	if node == nil {
+		if err != nil {
+			return err
+		}
+		return errors.New("Error getting node")
+	}
+
+	key, err := p.ensureSSHKey(p.config.Sshkey)
+	if err != nil {
+		return err
+	}
+
+	service := services.GetHardwareServerService(p.client)
+	config := datatypes.Container_Hardware_Server_Configuration{}
+	config.SshKeyIds = []int{*key}
+	_, err = service.Id(*node.Id).ReloadOperatingSystem(sl.String("FORCE"), &config)
+
+	if err != nil {
+		return err
+	}
+
+	// wait for node to switch to something not available
+	// if we don't delay here then the next reconcile will see an available node and mark is a ready
+	timeout := time.After(time.Minute) // Set the timeout to 1 minute
+	for {
+		select {
+		case <-timeout:
+			return nil
+		default:
+			node, _ := p.getNodeByName(id)
+			if node != nil && *node.HardwareStatus.Status != "ACTIVE" {
+				// Alls good, we can return
+				return nil
+			}
+			time.Sleep(time.Second * 3)
+		}
+	}
```

Comment:
Also in this case I don't think it's a good design to block the main thread for 1 minute. As a rule of thumb the provider methods should not be blocking, and rely to the `Completed` methods to ensure the operation is effectively done or not

Reply:
I'm not happy about this either but I'm not sure what else I can do.  Nodes stay in the "ACTIVE" state for a few seconds after running ReloadOperatingSystem, if we don't wait for the change to happen then Acquirecomplete will immediately move the CIR to "available" as the node in ibmcloud is ACTIVE, when infact it just hadn't started reloading yet.

Reply:
Would be possible to tag the resource somehow (ie metadata) before invoking `ReloadOperatingSystem` with a timestamp and then use it in the `CleanCompleted`? Ie the clean completed could wait as well at least 1 minute by looking at the timestamp, in a non blocking way

Reply:
tags don't work either they only have names not values, I think I found a way, testing it at the moment

Reply:
(a potential alternative could be to use the cir resource to store such timestamp, even though I'd prefer not)